apps/react/src/custom-hooks/useStateBuffered.ts

```diff
@@ -0,0 +1,57 @@
+import { useState } from "react";
+import { measureCallsPerSecond } from "../utils/logging";
+
+type StateTransformer<S> = (s: S) => S;
+type StateOrStateTransformer<S> = S | StateTransformer<S>;
+
+const STATE_CHANGE_EVENT_NAME = "stateChange";
+
+class StateBuffer<S> {
+  private state: S;
+  private isFrameRunning = false;
+  private pubSub = new EventTarget();
+
+  constructor(initialState: S) {
+    this.state = initialState;
+  }
+
+  updateState(stateOrFn: StateOrStateTransformer<S>) {
+    if (stateOrFn instanceof Function) {
+      this.state = stateOrFn(this.state);
+    } else {
+      this.state = stateOrFn;
+    }
+
+    this.runBuffer();
+  }
+
+  subscribe(cb: (s: S) => void) {
+    const handler = () => cb(this.state);
+    this.pubSub.addEventListener(STATE_CHANGE_EVENT_NAME, handler);
+    return () => this.pubSub.removeEventListener(STATE_CHANGE_EVENT_NAME, handler);
+  }
+
+  private publishState() {
+    this.pubSub.dispatchEvent(new CustomEvent(STATE_CHANGE_EVENT_NAME));
+  }
+
+  private runBuffer() {
+    if (!this.isFrameRunning) {
+      this.isFrameRunning = true;
+      requestAnimationFrame(() => {
+        this.publishState();
+        this.isFrameRunning = false;
+      });
+    }
+  }
+}
+
+export function useStateBuffered<S>(initialState: S): [S, (stateOrFn: StateOrStateTransformer<S>) => void] {
+  const [state, setState] = useState(initialState);
+  const buffer = new StateBuffer(initialState);
+  buffer.subscribe((state) => {
+    measureCallsPerSecond("setStateBufferred");
+    setState(state);
+  });
+  return [state, (stateOrFn: StateOrStateTransformer<S>) => buffer.updateState(stateOrFn)];
+}
```

Comment:
I think the pubsub is not needed and the hook could be simplified, one less memory leak to worry about: 

```suggestion
export function useStateBuffered<S>(initialState: S): [S, (stateOrFn: StateOrStateTransformer<S>) => void] {
  let isFrameRunning = false;
  let _state: S = initialState;
  const [state, setState] = useState(initialState);
  const updateState = (stateOrFn: StateOrStateTransformer<S>) => {
    if (stateOrFn instanceof Function) {
      _state = stateOrFn(_state);
    } else {
      _state = stateOrFn;
    }

    if (!isFrameRunning) {
      isFrameRunning = true;
      requestAnimationFrame(() => {
        setState(_state);
        isFrameRunning = false;
      });
    }
  }

  return [state, updateState];
}
```

Reply:
You do need the buffer.
However, we were both wrong - i forgot to put the buffer in a `useState` hook.
We broke the first rule of hooks - don't put local (stateful) variables in hooks !
You can't do stuff like this within a react hook:
```js
let _state: S = initialState;
// Or
let isFrameRunning = false;
```

because react hooks are called many times on each re-render.
These variables would just be reset to their original values.

You have to wrap them in useStates and then use the useEffect to initialize them.
I need the buffer thing so i can use a third party object to hold state without triggering reacts re-render.
Like so:

```typescript
export function useStateBuffered<S>(initialState: S): [S, (stateOrFn: StateOrStateTransformer<S>) => void] {
  const [state, setState] = useState(initialState);
  const [buffer] = useState(() => new StateBuffer(initialState))); // creates buffer only once for component. before it was re-initialised many times
  useEffect(() => buffer.subscribe((state) => { // subscribe within a use-effect to so it will only subscribe once. we need the buffer because you can't put state in local variables in a hook. You have to have some 3rd party state holder. like a store or this simple object i created.
    measureCallsPerSecond("setStateBufferred");
    setState(state);
  }), []);
  return [state, (stateOrFn: StateOrStateTransformer<S>) => buffer.updateState(stateOrFn)];
}
```

Reply:
What do you need the pubsub for? I don't understand what value it brings

I think refs fit this use case better, they persist data between renders and when you change them do not trigger a re-render